File: kinotic-management-api/src/main/java/org/kinotic/management/api/model/workload/WorkloadStatus.java
package org.kinotic.management.api.model.workload;

import lombok.Getter;

/**
 * Represents the current status of a {@link Workload}. A run moves through these in order and only
 * forward: pending, starting, running, stopping, then ended, where the two ended statuses share one
 * rank since a run ends once.
 */
public enum WorkloadStatus {
    PENDING(0),
    STARTING(1),
    RUNNING(2),
    STOPPING(3),
    STOPPED(4),
    FAILED(4);

    /**
     * Where this status falls in a run: a status of a lower rank comes earlier, and the two ended
     * statuses share the highest.
     */
    @Getter
    private final int rank;

    WorkloadStatus(int rank) {
        this.rank = rank;
    }

    /**
     * True when this run of the workload has ended — the guest is no longer executing and the
     * node removes its VM.
     */
    public boolean isComplete() {
        return this == STOPPED || this == FAILED;
    }

    /**
     * True while the run holds a VM on its node: started and not yet ended. A stop the node has not
     * answered counts, since the VM may still be there.
     */
    public boolean isOpen() {
        return this == STARTING || this == RUNNING || this == STOPPING;
    }

    /**
     * True when this status comes later in a run than {@code other}, so a report of it moves the run
     * forward.
     */
    public boolean isAfter(WorkloadStatus other) {
        return rank > other.rank;
    }
}

File: kinotic-js/workspace/packages/management-api/src/api/model/workload/WorkloadStatus.ts

/**
 * The current status of a Workload.
 */
export enum WorkloadStatus {
    /** Accepted, not yet placed on a node. */
    PENDING = 'PENDING',
    /** Placed on a node, which is bringing the VM up. */
    STARTING = 'STARTING',
    /** The guest is executing. */
    RUNNING = 'RUNNING',
    /** A stop was requested and the node is shutting the guest down. */
    STOPPING = 'STOPPING',
    /**
     * The run was ended by a stop request, or a detached run's guest exited on its own with
     * code 0. The VM is dormant, unless autoRemove discarded it.
     */
    STOPPED = 'STOPPED',
    /**
     * A non-detached run's guest ran to its end and exited with code 0. The VM is dormant,
     * unless autoRemove discarded it.
     */
    COMPLETED = 'COMPLETED',
    /**
     * The VM could not be started, its guest exited on its own with a non-zero code, or its
     * node went offline while it ran.
     */
    FAILED = 'FAILED'
}
